### PR TITLE
Add FX10946 [v103] Tabs Integration Telemetry

### PR DESCRIPTION
**Pull request by lougeniaC64**
This PR is based on #10553 and should only be merged after that PR has been merged. This PR creates telemetry to help track the sync success metrics of the appservices tabs component and the native clients engine.

### CODE REVIEW

**Comment by lougeniaC64** [comment]
All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

    What questions will you answer with this data?

  - This will help us to separately gauge the success rates of the native iOS tabs component and the application services tabs component as well as the client engine's sync success rates.

**Comment by lougeniaC64** [comment]
@eliserichards Ready for data review! Let me know if I need to make any changes.

@nishant2718 This PR is includes the telemetry changes for the tabs integration and shouldn't be merged before #10553. Because this PR is based on that tabs integration PR, it will decrease to the changes in [this commit](https://github.com/mozilla-mobile/firefox-ios/pull/10915/commits/0ff4db389d06490acc7c03e0727107c2e75acd61) once the tabs integration has been merged.

**Comment by eliserichards** [comment]
Just a small suggestion, then I'm happy to data-review+: 

> 7. How long will this data be collected?
> * This collection has expiry '2023-01-01'

I would recommend changing this to '2023-01-07' or something similar. Things can get lost at the end of the year and I don't want these metrics to accidentally expire while we are all on break 😋

**Comment by eliserichards** [comment]
Whoops, looks like I don't have commit rights for this branch. Hopefully this will make your life easier :) 

Updates dates for `metrics.yaml`: [patch1.patch.zip](https://github.com/mozilla-mobile/firefox-ios/files/8903713/patch1.patch.zip)

**Comment by eliserichards** [comment]
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
>     1. What questions will you answer with this data?
> 
> 
>     * This will help us to separately gauge the success rates of the native iOS tabs component and the application services tabs component as well as the client engine's sync success rates.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
>     * We need to answer these questions to ensure that replacing the native tabs component with the application services tabs component doesn't result in degraded syncing for users. We also want to start tracking the sync success rates of the client engine ahead of its upcoming replacement.
> 
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
>     * There are no other alternatives.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * No.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.    
>        Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> 
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug
> rust_tabs_sync.uid 	The user's hashed Firefox Account ID. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> rust_tabs_sync.incoming 	Records incoming tabs record counts. applied is the number of incoming records that were successfully stored or updated in the local database. failed_to_apply is the number of records that were ignored due to errors. reconciled is the number of merged records. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> rust_tabs_sync.outgoing 	Records outgoing tabs record counts. uploaded is the number of records that were successfully sent to the server. failed_to_upload is the number of records that weren't uploaded, and will be retried on the next sync. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> rust_tabs_sync.failure_reason 	Records why the tabs sync failed. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> clients_sync.uid 	The user's hashed Firefox Account ID. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> clients_sync.incoming 	Records incoming clients record counts. applied is the number of incoming records that were successfully stored or updated in the local database. failed_to_apply is the number of records that were ignored due to errors. reconciled is the number of merged records. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> clients_sync.outgoing 	Records outgoing clients record counts. uploaded is the number of records that were successfully sent to the server. failed_to_upload is the number of records that weren't uploaded, and will be retried on the next sync. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> clients_sync.failure_reason 	Records why the clients sync failed. 	interaction 	https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/boards/211?modal=detail&selectedIssue=SYNC-3170
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
>     * This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org/).
> 
> 
>     7. How long will this data be collected?
> 
> 
>     * This collection has expiry '2023-01-01'
> 
> 
>     8. What populations will you measure?
> 
> 
>     * All channels, countries, and locales. No filters.
> 
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
>     * These collections are Glean. The opt-out can be found in the product's preferences.
> 
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
>     * This data will be analyzed through existing and existing Redash dashboard.
> 
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
>     * The results will be Mozilla internal.
> 
> 
>     12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
> 
> 
>     * No.

# Data Review Form (to be filled by Data Stewards)

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

    * Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/)

2) Is there a control mechanism that allows the user to turn the data collection on and off? (Note, for data collection not needed for security purposes, Mozilla provides such a control mechanism) Provide details as to the control mechanism available.

    * Yes, through the "Send Usage Data" preference in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

    * No, the data will be collected until January 7, 2023 with the option to remove/renew at that time.

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

    * Category 2 - interaction data

5) Is the data collection request for default-on or default-off?

    * Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

    * No

7) Is the data collection covered by the existing Firefox privacy notice? 

    * Yes

8) Does the data collection use a third-party collection tool? **If yes, escalate to legal.**

    * No

## Result

data-review+

File: Providers/Profile.swift
Comment (electricRGB) [inline Providers/Profile.swift:529]:
```suggestion
            guard let clientRemoteTabs = tabsResult.successValue else { return deferMaybe([]) }
```
nit

File: Providers/Profile.swift
Comment (electricRGB) [inline Providers/Profile.swift:525]:
```suggestion
            if let tabsError = tabsResult.failureValue { return deferMaybe(tabsError) }
```
nit

File: Providers/Profile.swift
Comment (electricRGB) [inline Providers/Profile.swift:534]:
```suggestion
                if let error = result.failureValue { return deferMaybe(error) }
```
nit

File: Providers/Profile.swift
Comment (electricRGB) [inline Providers/Profile.swift:538]:
```suggestion
                guard let clients = result.successValue else { return deferMaybe([]) }
```

File: Storage/Rust/RustRemoteTabs.swift
Comment (electricRGB) [inline Storage/Rust/RustRemoteTabs.swift:9]: We have a `Loggable` protocol to replace these `private let log....` danglies at the top of files. It'd be great if we could replace it with the new standard way of doing things. :)

File: Storage/Rust/RustRemoteTabs.swift
Comment (electricRGB) [inline Storage/Rust/RustRemoteTabs.swift:15]:
```suggestion
        guard let url = Foundation.URL(string: self.urlHistory[0]) else { return nil }
```
nit

File: Providers/Profile.swift
Comment (electricRGB) [inline Providers/Profile.swift:1076]: I know you didn't create it, but can we rename this to something more legible? Single letter variable names are not informative.

**Review by electricRGB — CHANGES_REQUESTED** [review]
Heya. Requesting just a few minor changes for standardization.

The nit's are up to you however. Generally we prefer small statements on one line, for future reference. There's a few places where lines are really long. Again, for future reference, if we can line break these and indent them appropriately - I'd have tried to fix them here, but GH's indentation is hot trash so I'd do more harm than good. I'd recommend going to Xcode preferences -> Text Editing and enabling "Page column guide at:" The recommended for the codebase is 120, but the closer you are to 90/100, the better for readability. :)

**Review by electricRGB — APPROVED** [review]

**Pull request merged**
